logstash-core-event-java/src/main/java/com/logstash/Javafier.java

```diff
@@ -1,44 +1,76 @@
 package com.logstash;
 
+import com.logstash.ext.JrubyTimestampExtLibrary;
+import org.joda.time.DateTime;
 import org.jruby.RubyArray;
-import org.jruby.RubyHash;
-import org.jruby.RubyString;
-import org.jruby.RubyObject;
+import org.jruby.RubyBignum;
 import org.jruby.RubyBoolean;
-import org.jruby.RubyArray;
+import org.jruby.RubyFixnum;
 import org.jruby.RubyFloat;
+import org.jruby.RubyHash;
 import org.jruby.RubyInteger;
 import org.jruby.RubyNil;
-import org.jruby.RubyBoolean;
-import org.jruby.RubyFixnum;
-import org.jruby.RubyTime;
+import org.jruby.RubyString;
 import org.jruby.RubySymbol;
-import org.jruby.RubyBignum;
+import org.jruby.RubyTime;
 import org.jruby.ext.bigdecimal.RubyBigDecimal;
-import com.logstash.ext.JrubyTimestampExtLibrary;
+import org.jruby.java.proxies.JavaProxy;
+import org.jruby.java.proxies.MapJavaProxy;
+import org.jruby.javasupport.JavaUtil;
 import org.jruby.runtime.builtin.IRubyObject;
-import java.math.BigDecimal;
-import org.joda.time.DateTime;
 
+import java.math.BigDecimal;
 import java.math.BigInteger;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 public class Javafier {
+    private static final String ERR_TEMPLATE = "Missing Ruby class handling for full class name=%s, simple name=%s";
+    private static final String PROXY_ERR_TEMPLATE = "Missing Ruby class handling for full class name=%s, simple name=%s, wrapped object=%s";
 
     private Javafier(){}
 
-    public static List<Object> deep(RubyArray a) {
+    public static List<Object> deep(IRubyObject[] a) {
         final ArrayList<Object> result = new ArrayList();
 
-        // TODO: (colin) investagate why .toJavaArrayUnsafe() which should be faster by avoiding copying produces nil values spec errors in arrays
-        for (IRubyObject o : a.toJavaArray()) {
+        for (IRubyObject o : a) {
             result.add(deep(o));
         }
         return result;
     }
 
+    public static List<Object> deep(RubyArray a) {
```

Comment:
err, at least it is not being tested. From my understanding `deep(IRubyObject)` will be called for these instances, since `RubyArray` might stay as generic as an IRubyObject